front/lib/api/assistant/generation.ts

```diff
@@ -145,6 +147,32 @@ export async function renderConversationForModel({
       return new Err(new Error(`Error tokenizing model message: ${err}`));
     }
   }
+  async function tokenSplit(
+    text: string,
+    model: { providerId: string; modelId: string },
+    splitAt: number
+  ): Promise<Result<string, Error>> {
+    try {
+      const coreAPI = new CoreAPI(logger);
+      const res = await coreAPI.tokenize({
+        text,
+        providerId: model.providerId,
+        modelId: model.modelId,
+      });
+      if (res.isErr()) {
+        return new Err(
+          new Error(`Error tokenizing model message: ${res.error.message}`)
+        );
+      }
+      const remainingText = res.value.tokens
+        .slice(0, splitAt)
+        .map(([, tokenText]) => tokenText)
+        .join("");
+      return new Ok(safeSubstring(remainingText, 0, remainingText.length));
```

Comment:
This thing can't recover if the last character is a half unicode no?

Reply:
Does it add more character in case of failure or does it remove them?

Reply:
it removes them (cf `safesubstring doc`) so it should be fine

Reply:
ok !